rfcs/text/0014_api_documentation.md

```diff
@@ -0,0 +1,439 @@
+
+
+```ts
+export interface PluginApi {
+  id: string;
+  serviceFolders?: readonly string[];
+  client: ScopeApi;
+  server: ScopeApi;
+  common: ScopeApi;
+}
+```
+
+## Manifest file changes
+
+**serviceFolders?: string[]**
+
+Used by the system to group services into sub-pages. Some plugins, like data and core, have such huge APIs they are very slow to contain in a single page, and they are less consummable by solution developers. The addition of an optional list of services folders will cause the system to automatically create a separate page with every API that is defined within that folder. The caveat is that core will need to define a manifest file in order to define its service folders...
+
+**Using a kibana.json file for core**
+
+For the purpose of API infrastructure, core is treated like any other plugin. This means it has to specify serviceFolders section inside a manifest file to be split into sub folders. There are other ways to tackle this - like a hard coded array just for the core folder, but I kept the logic as similar to the other plugins as possible.
+
+**teamOwner: string**
+
+Team owner can be determined via github CODEOWNERS file, but we want to encourage single team ownership per plugin. Requiring a team owner string in the manifest file will help with this and will allow the API doc system to manually add a section to every page that has a link to the team owner. Additional ideas are teamSlackChannel or teamEmail for further contact. 
+
+**summary: string**
+
+A brief description of the plugin can then be displayed in the automatically generated API documentation.
+
+# Future features
+
+## Indexing stats
+
+Can we index statistics about our API as part of this system? For example, I'm dumping information about which api declarations are missing comments in the console.
+
+## Longer term approach to "plugin service folders"
+
+Using sub folders is a short term plan. A long term plan hasn't been established yet, but it should fit in with our folder structure hierarchy goals, along with
+any support we have for sharing services among a related set of plugins, that are not exposed as part of the public API.
+# Recommendations for writing comments
+
+## @link comments for the referenced type
+
+Core has a pattern of writing comments like this:
+
+```ts
+  /** {@link IUiSettingsClient} */
+  uiSettings: IUiSettingsClient;
+```
+
+I don't see the value in this. In the IDE, I can click on the IUiSettingsClient type and get directed there, and in the API doc system, the
+type will already be clickable. This ends up with a weird looking API:
+
+![image](../images/repeat_type_links.png)
+
+The plan is to make @link comments work like links, which means this is unneccessary information.
+
+I propose we avoid this kind of pattern.
+
+## Export every referenced type
+
+The docs system handles broken link warnings but to avoid breaking the ci, I suggest we turn this off initially. This will mean however,
```

Comment:
nit
```suggestion
The docs system handles broken link warnings but to avoid breaking CI, I suggest we turn this off initially. However, this will mean
```